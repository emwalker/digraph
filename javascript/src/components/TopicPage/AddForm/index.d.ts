import { AddForm_topic$data as Topic } from '__generated__/AddForm_topic.graphql';
import { AddForm_viewer$data as Viewer } from '__generated__/AddForm_viewer.graphql';
import './index.css';
declare type Props = {
    topic: Topic;
    viewer: Viewer;
};
declare const _default: import("react-relay").Container<Omit<Props, "relay">>;
export default _default;
