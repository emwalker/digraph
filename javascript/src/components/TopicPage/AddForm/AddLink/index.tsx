import React, { Component, KeyboardEvent, FormEvent } from 'react'
import { graphql, createFragmentContainer, RelayProp, DeclarativeMutationConfig } from 'react-relay'

import upsertLinkMutation, { Input } from 'mutations/upsertLinkMutation'
import { AddLink_viewer as ViewerType } from '__generated__/AddLink_viewer.graphql'
import { AddLink_topic as TopicType } from '__generated__/AddLink_topic.graphql'

type RepositoryType = ViewerType['selectedRepository']

const tooltip = 'Add a link to this topic.\n'
  + 'Press "Return" to submit the new link.'

type Props = {
  disabled?: boolean,
  relay: RelayProp,
  topic: TopicType,
  viewer: ViewerType,
}

type State = {
  url: string,
}

class AddLink extends Component<Props, State> {
  static defaultProps = {
    disabled: false,
  }

  constructor(props: Props) {
    super(props)
    this.state = {
      url: '',
    }
  }

  onKeyPress = (event: KeyboardEvent<HTMLInputElement>) => {
    if (event.key === 'Enter') this.createLink()
  }

  get selectedRepo(): RepositoryType {
    return this.props.viewer.selectedRepository
  }

  get orgLogin() {
    const repo = this.selectedRepo
    if (!repo) return null
    return repo.organization.login
  }

  get relayConfigs(): DeclarativeMutationConfig[] {
    return [{
      type: 'RANGE_ADD',
      parentID: this.props.topic.id,
      connectionInfo: [{
        key: 'Topic_links',
        rangeBehavior: 'prepend',
      }],
      edgeName: 'linkEdge',
    }]
  }

  updateUrl = (event: FormEvent<HTMLInputElement>) => {
    this.setState({ url: event.currentTarget.value })
  }

  createLink() {
    const repo = this.selectedRepo
    const repoName = repo ? repo.name : null
    const { orgLogin } = this

    if (!repoName) return
    if (!orgLogin) return

    const input: Input = {
      addParentTopicIds: [this.props.topic.id],
      organizationLogin: orgLogin,
      repositoryName: repoName,
      url: this.state.url,
    }

    upsertLinkMutation(
      this.props.relay.environment,
      input,
      {
        configs: this.relayConfigs,
      },
    )
    this.setState({ url: '' })
  }

  render = () => (
    <dl className="form-group">
      <dt>
        <span
          className="tooltipped tooltipped-ne"
          aria-label={tooltip}
        >
          <label htmlFor="create-link-url">Add link</label>
        </span>
      </dt>
      <dd>
        <input
          className="form-control test-link-url input-sm"
          disabled={this.props.disabled}
          id="create-link-url"
          onChange={this.updateUrl}
          onKeyPress={this.onKeyPress}
          placeholder="Url"
          type="url"
          value={this.state.url}
        />
      </dd>
    </dl>
  )
}

export default createFragmentContainer(AddLink, {
  viewer: graphql`
    fragment AddLink_viewer on User {
      selectedRepository {
        id
        name

        organization {
          login
        }
      }
    }
  `,
  topic: graphql`
    fragment AddLink_topic on Topic {
      id
    }
  `,
})
