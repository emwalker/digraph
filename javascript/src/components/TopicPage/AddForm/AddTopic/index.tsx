import React, { Component, KeyboardEvent, FormEvent } from 'react'
import { graphql, createFragmentContainer, RelayProp, DeclarativeMutationConfig } from 'react-relay'

import upsertTopicMutation from 'mutations/upsertTopicMutation'
import { AddTopic_viewer as Viewer } from '__generated__/AddTopic_viewer.graphql'
import { AddTopic_topic as Topic } from '__generated__/AddTopic_topic.graphql'

type RepositoryType = Viewer['selectedRepository']

const tooltipText = 'Add a subtopic to this topic. You can click "Edit"\n'
  + 'afterwards if it also belongs under another topic.\n'
  + 'Press "Return" to submit the new topic.'

type Props = {
  disabled?: boolean,
  relay: RelayProp,
  topic: Topic,
  viewer: Viewer,
}

type State = {
  name: string,
}

class AddTopic extends Component<Props, State> {
  static defaultProps = {
    disabled: true,
  }

  constructor(props: Props) {
    super(props)
    this.state = {
      name: '',
    }
  }

  onKeyPress = (event: KeyboardEvent<HTMLInputElement>) => {
    if (event.key === 'Enter') this.createTopic()
  }

  get selectedRepo(): RepositoryType {
    return this.props.viewer.selectedRepository
  }

  get orgLogin(): string | null {
    const repo = this.selectedRepo
    if (!repo) return null
    return repo.organization.login
  }

  get relayConfigs(): DeclarativeMutationConfig[] {
    return [{
      type: 'RANGE_ADD',
      parentID: this.props.topic.id,
      connectionInfo: [{
        key: 'Topic_childTopics',
        rangeBehavior: 'prepend',
      }],
      edgeName: 'topicEdge',
    }]
  }

  updateName = (event: FormEvent<HTMLInputElement>) => {
    this.setState({ name: event.currentTarget.value })
  }

  createTopic() {
    const repo = this.selectedRepo
    if (!repo || !this.orgLogin) {
      // eslint-disable-next-line no-console
      console.log('Missing org name or repo, unable to create topic')
      return
    }

    upsertTopicMutation(
      this.props.relay.environment,
      {
        name: this.state.name,
        repositoryName: repo.name,
        organizationLogin: this.orgLogin,
        topicIds: [this.props.topic.id],
      },
      {
        configs: this.relayConfigs,
      },
    )
    this.setState({ name: '' })
  }

  render = () => (
    <dl className="form-group">
      <dt>
        <span
          className="tooltipped tooltipped-ne"
          aria-label={tooltipText}
        >
          <label htmlFor="create-topic-name">Add subtopic</label>
        </span>
      </dt>
      <dd>
        <input
          className="form-control test-topic-name input-sm"
          disabled={this.props.disabled}
          id="create-topic-name"
          onChange={this.updateName}
          onKeyPress={this.onKeyPress}
          placeholder="Name or description"
          value={this.state.name}
        />
      </dd>
    </dl>
  )
}

export default createFragmentContainer(AddTopic, {
  viewer: graphql`
    fragment AddTopic_viewer on User {
      selectedRepository {
        name
        organization {
          login
        }
      }
    }
  `,
  topic: graphql`
    fragment AddTopic_topic on Topic {
      id
    }
  `,
})
