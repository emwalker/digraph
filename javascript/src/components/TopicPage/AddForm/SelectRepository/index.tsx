import React, { Component, FormEvent } from 'react'
import { createFragmentContainer, graphql, RelayProp } from 'react-relay'

import { EdgeTypeOf } from 'components/types'
import selectRepositoryMutation, { Input } from 'mutations/selectRepositoryMutation'
import {
  SelectRepository_viewer as ViewerType,
} from '__generated__/SelectRepository_viewer.graphql'

type EdgeType = EdgeTypeOf<ViewerType['repositories']>

type Props = {
  relay: RelayProp,
  viewer: ViewerType,
}

class SelectRepository extends Component<Props> {
  onChange = (event: FormEvent<HTMLSelectElement>) => {
    const repositoryId = event.currentTarget.value
    const input: Input = {
      repositoryId: repositoryId === 'placeholder' ? null : repositoryId,
    }
    selectRepositoryMutation(this.props.relay.environment, input)
  }

  get repositoryEdges() {
    const { repositories } = this.props.viewer
    return repositories?.edges || []
  }

  get selectedId() {
    const repo = this.props.viewer.selectedRepository
    return repo ? repo.id : null
  }

  renderOption = (edge: EdgeType) => (
    edge && (
      <option
        key={edge.node.fullName}
        value={edge.node.id || undefined}
      >
        {edge.node.fullName}
      </option>
    )
  )

  render = () => (
    <dl className="form-group">
      <dt>
        <label htmlFor="select-repo">New links and topics added to</label>
      </dt>
      <dd>
        <select
          id="select-repo"
          className="form-select"
          aria-label="Repository"
          style={{ width: '100%' }}
          defaultValue={this.selectedId || undefined}
          onChange={this.onChange}
        >
          <option key="0" value="placeholder">Select a repository</option>
          {this.repositoryEdges.map(this.renderOption)}
        </select>
      </dd>
    </dl>
  )
}

export default createFragmentContainer(SelectRepository, {
  viewer: graphql`
    fragment SelectRepository_viewer on User {
      selectedRepository {
        id
        isPrivate
      }

      repositories(first: 100) {
        edges {
          isSelected

          node {
            fullName
            id
          }
        }
      }
    }
  `,
})
