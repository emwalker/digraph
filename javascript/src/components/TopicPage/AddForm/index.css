.private-repo {
  color: #735c0f;
  background-color: #fffbdd;
}
