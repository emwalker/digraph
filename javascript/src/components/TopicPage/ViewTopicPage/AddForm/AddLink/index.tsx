import React, {
  FormEvent, useState, useCallback, KeyboardEvent,
} from 'react'
import { graphql, useFragment } from 'react-relay'

import { makeUpsertLinkCallback } from 'mutations/upsertLinkMutation'
import {
  AddLink_viewer$key,
} from '__generated__/AddLink_viewer.graphql'
import {
  AddLink_topic$key,
} from '__generated__/AddLink_topic.graphql'

const tooltip = 'Add a link to this topic.\n'
  + 'Press "Return" to submit the new link.'

type Props = {
  disabled?: boolean,
  topic: AddLink_topic$key,
  viewer: AddLink_viewer$key,
}

const viewerFragment = graphql`
  fragment AddLink_viewer on User {
    selectedRepoId
  }
`

const topicFragment = graphql`
  fragment AddLink_topic on Topic {
    id
  }
`

export default function AddLink(props: Props) {
  const viewer = useFragment(viewerFragment, props.viewer)
  const topic = useFragment(topicFragment, props.topic)
  const [url, setUrl] = useState('')
  const selectedRepoId = viewer.selectedRepoId
  const upsertLink = makeUpsertLinkCallback({ url, setUrl, topicId: topic.id, selectedRepoId })

  const onKeyPress = useCallback((event: KeyboardEvent<HTMLInputElement>) => {
    if (event.key !== 'Enter') return
    upsertLink()
  }, [upsertLink])

  const onChange = useCallback((event: FormEvent<HTMLInputElement>) => {
    setUrl(event.currentTarget.value)
  }, [setUrl])

  return (
    <dl className="form-group">
      <dt>
        <span
          className="tooltipped tooltipped-ne"
          aria-label={tooltip}
        >
          <label htmlFor="create-link-url">Add link</label>
        </span>
      </dt>
      <dd>
        <input
          className="form-control input-sm"
          disabled={props.disabled}
          data-testid="link-url-input"
          id="create-link-url"
          onChange={onChange}
          onKeyPress={onKeyPress}
          placeholder="Url"
          type="url"
          value={url}
        />
      </dd>
    </dl>
  )
}
