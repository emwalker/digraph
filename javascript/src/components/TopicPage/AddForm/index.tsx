import React, { Component } from 'react'
import { createFragmentContainer, graphql } from 'react-relay'
import classNames from 'classnames'

import { AddForm_topic as Topic } from '__generated__/AddForm_topic.graphql'
import { AddForm_viewer as Viewer } from '__generated__/AddForm_viewer.graphql'
import AddTopic from './AddTopic'
import AddLink from './AddLink'
import SelectRepository from './SelectRepository'
import './index.css'

type Props = {
  topic: Topic,
  viewer: Viewer,
}

class AddForm extends Component<Props> {
  get className(): string {
    return classNames(
      'border',
      'rounded-1',
      'px-md-2',
      'px-3',
      'mt-3',
    )
  }

  get isPrivateRepo(): boolean {
    const repo = this.props.viewer.selectedRepository
    if (!repo) return false
    return repo.isPrivate
  }

  get repoSelected(): boolean {
    return !!this.props.viewer.selectedRepository
  }

  renderInputFields = () => (
    <>
      <AddTopic
        disabled={!this.repoSelected}
        topic={this.props.topic}
        viewer={this.props.viewer}
      />
      <AddLink
        disabled={!this.repoSelected}
        topic={this.props.topic}
        viewer={this.props.viewer}
      />
    </>
  )

  get selectRepositoryStyle(): Object {
    const backgroundColor = this.isPrivateRepo ?
      this.props.viewer.selectedRepository?.displayColor :
      'transparent'
    return { backgroundColor }
  }

  render = () => (
    <form className={this.className} style={this.selectRepositoryStyle}>
      <SelectRepository
        viewer={this.props.viewer}
      />
      {this.repoSelected && this.renderInputFields()}
    </form>
  )
}

export default createFragmentContainer(AddForm, {
  viewer: graphql`
    fragment AddForm_viewer on User {
      selectedRepository {
        isPrivate
        displayColor
      }

      ...AddLink_viewer
      ...AddTopic_viewer
      ...SelectRepository_viewer
      ...SelectedRepo_viewer
    }
  `,
  topic: graphql`
    fragment AddForm_topic on Topic {
      ...AddLink_topic
      ...AddTopic_topic
    }
  `,
})
