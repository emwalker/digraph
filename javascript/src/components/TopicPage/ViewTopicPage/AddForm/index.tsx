import React from 'react'
import { graphql, useFragment } from 'react-relay'

import {
  AddForm_topic$key, AddForm_topic$data as TopicType,
} from '__generated__/AddForm_topic.graphql'
import {
  AddForm_viewer$key, AddForm_viewer$data as ViewerType,
} from '__generated__/AddForm_viewer.graphql'
import AddTopic from './AddTopic'
import AddLink from './AddLink'
import SelectRepository from './SelectRepository'
import './index.css'
import Blankslate from 'components/ui/Blankslate'

type Props = {
  topic: AddForm_topic$key,
  viewer: AddForm_viewer$key,
}

const viewerFragment = graphql`
  fragment AddForm_viewer on User {
    selectedRepoId

    selectedRepo {
      isPrivate
      displayColor
    }

    ...AddLink_viewer
    ...AddTopic_viewer
    ...SelectRepository_viewer
    ...SelectedRepo_viewer
  }
`

const topicFragment = graphql`
  fragment AddForm_topic on Topic {
    ...AddLink_parentTopic
    ...AddTopic_parentTopic

    repoTopics {
      repoId
    }
  }
`

type InnerAddFormProps = {
  isPrivateRepo: boolean,
  viewer: ViewerType,
  topic: TopicType,
}

function InnerAddForm({ isPrivateRepo, viewer, topic }: InnerAddFormProps) {
  const selectedRepoId = viewer.selectedRepoId
  const hasSelectedRepo = !!viewer.selectedRepo
  const topicInSelectedRepo = !!topic.repoTopics
    .find((repoTopic) => repoTopic.repoId == selectedRepoId)
  const canUpsert = hasSelectedRepo && (isPrivateRepo || topicInSelectedRepo)

  if (!selectedRepoId) return null

  if (canUpsert) {
    return (
      <>
        <AddTopic disabled={!hasSelectedRepo} parentTopic={topic} viewer={viewer} />
        <AddLink disabled={!hasSelectedRepo} parentTopic={topic} viewer={viewer} />
      </>
    )
  }

  return (
    <div data-testid="upserts-disabled" >
      <Blankslate>
        <p>This is a private topic, and in order to protect confidentiality,
          subtopics and links cannot be added to it by way of a public repo</p>
      </Blankslate>
    </div >
  )
}

export default function AddForm(props: Props) {
  const viewer = useFragment(viewerFragment, props.viewer)
  const topic = useFragment(topicFragment, props.topic)
  const isPrivateRepo = !!viewer.selectedRepo?.isPrivate

  const selectRepositoryStyle = {
    backgroundColor: isPrivateRepo ?
      viewer.selectedRepo?.displayColor :
      'transparent',
  }

  return (
    <form className="border rounded-1 px-md-2 px-3 mt-3" style={selectRepositoryStyle}>
      <SelectRepository viewer={viewer} />
      <InnerAddForm isPrivateRepo={isPrivateRepo} topic={topic} viewer={viewer} />
    </form>
  )
}
