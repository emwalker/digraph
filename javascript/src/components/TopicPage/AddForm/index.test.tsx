import React from 'react'
import { shallow } from 'enzyme'

import AddForm from './index'

jest.mock('react-relay', () => ({
  createFragmentContainer: (component: any) => component,
}))

const defaultViewer = {
  selectedRepository: {
    name: null,
    organization: {
      login: 'gnusto',
    },
  },
}

const props = {
  topic: {
    id: '1234',
  },
  viewer: defaultViewer,
}

describe('<AddForm />', () => {
  // @ts-expect-error
  const wrapper = shallow(<AddForm {...props} />)

  it('renders', () => {
    expect(wrapper).toMatchSnapshot()
  })

  describe('selecting a repository', () => {
    const addLink = () => wrapper.find('AddLink')
    const addTopic = () => wrapper.find('AddTopic')

    describe('when a repository has been selected', () => {
      beforeEach(() => {
        wrapper.setProps({ viewer: defaultViewer })
      })

      it('enables the input fields', () => {
        expect(addLink().exists()).toBeTruthy()
        expect(addTopic().exists()).toBeTruthy()
      })
    })

    describe('when a repository has not yet been selected', () => {
      beforeEach(() => {
        wrapper.setProps({ viewer: { selectedRepository: null } })
      })

      it('disables the input fields', () => {
        expect(addLink().exists()).toBeFalsy()
        expect(addTopic().exists()).toBeFalsy()
      })
    })
  })
})
