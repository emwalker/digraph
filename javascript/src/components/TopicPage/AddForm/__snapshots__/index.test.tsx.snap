// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddForm /> renders 1`] = `
<form
  className="border rounded-1 px-md-2 px-3 mt-3"
>
  <SelectRepository
    viewer={
      Object {
        "selectedRepository": Object {
          "name": null,
          "organization": Object {
            "login": "gnusto",
          },
        },
      }
    }
  />
  <AddTopic
    disabled={false}
    topic={
      Object {
        "id": "1234",
      }
    }
    viewer={
      Object {
        "selectedRepository": Object {
          "name": null,
          "organization": Object {
            "login": "gnusto",
          },
        },
      }
    }
  />
  <AddLink
    disabled={false}
    topic={
      Object {
        "id": "1234",
      }
    }
    viewer={
      Object {
        "selectedRepository": Object {
          "name": null,
          "organization": Object {
            "login": "gnusto",
          },
        },
      }
    }
  />
</form>
`;
